File: ios/StillMoment/Presentation/Views/GuidedMeditations/TrimEditor/TrimPlayheadGrabber.swift
//
//  TrimPlayheadGrabber.swift
//  Still Moment
//
//  Presentation Layer — sage playhead grabber in the waveform's upper zone (shared-108).
//

import SwiftUI

/// The sage playhead grabber, rendered as a full-size overlay over the waveform and
/// positioned in the upper (playhead) touch zone. The full-height playhead line is
/// drawn by `TrimWaveformView`; this is only the grip.
///
/// Purely visual — the actual dragging is resolved geometrically by the single track
/// gesture in `TrimWaveformSection` (the upper 45 % of the waveform belongs to the
/// playhead). Exposed as an adjustable accessibility element (±1 s seeks), also while
/// the playhead is outside the zoom window and the grip is hidden.
struct TrimPlayheadGrabber: View {
    // MARK: Internal

    let playheadTime: TimeInterval
    /// Visible time window (zoom, shared-108) — an off-window playhead shows no grip
    /// (never glued to the edge).
    let window: ClosedRange<TimeInterval>
    let trackWidth: CGFloat
    let isDragging: Bool
    let onNudge: (TimeInterval) -> Void

    var body: some View {
        ZStack(alignment: .topLeading) {
            if self.trackWidth > 0, TrimGeometry.isTime(self.playheadTime, inWindow: self.window) {
                self.grabber
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .allowsHitTesting(false)
        .accessibilityElement()
        .accessibilityLabel(Text("trim_editor.a11y.playhead"))
        .accessibilityValue(Text(EditSheetState.formatTime(self.playheadTime)))
        .accessibilityAdjustableAction { direction in
            switch direction {
            case .increment: self.onNudge(1)
            case .decrement: self.onNudge(-1)
            @unknown default: break
            }
        }
    }

    // MARK: Private

    @Environment(\.themeColors)
    private var theme

    private static let grabberSize = CGSize(width: 32, height: 20)
    private static let tipSize = CGSize(width: 10, height: 6)

    private var headX: CGFloat {
        TrimGeometry.x(for: self.playheadTime, window: self.window, width: self.trackWidth)
    }

    /// Vertical center of the playhead (upper) touch zone.
    private var zoneCenterY: CGFloat {
        TrimWaveformView.height * TrimHitTesting.verticalSplit / 2
    }

    private var grabber: some View {
        VStack(spacing: -1) {
            RoundedRectangle(cornerRadius: 7)
                .fill(
                    LinearGradient(
                        colors: [self.theme.playheadAccentHi, self.theme.playheadAccent],
                        startPoint: .top,
                        endPoint: .bottom
                    )
                )
                .frame(width: Self.grabberSize.width, height: Self.grabberSize.height)
                .overlay(
                    HStack(spacing: 3) {
                        self.grip
                        self.grip
                    }
                )
                .shadow(color: self.theme.cardShadow, radius: 4, y: 2)
            Triangle()
                .fill(self.theme.playheadAccent)
                .frame(width: Self.tipSize.width, height: Self.tipSize.height)
        }
        .overlay(alignment: .top) {
            if self.isDragging {
                TrimTimeBubble(
                    time: self.playheadTime,
                    anchorX: self.headX,
                    trackWidth: self.trackWidth,
                    background: self.theme.playheadAccentHi,
                    textColor: self.theme.textOnPlayhead
                )
                .offset(y: -30)
            }
        }
        .offset(
            x: self.headX - Self.grabberSize.width / 2,
            y: self.zoneCenterY - (Self.grabberSize.height + Self.tipSize.height) / 2
        )
    }

    private var grip: some View {
        Capsule()
            .fill(self.theme.textOnPlayhead.opacity(0.55))
            .frame(width: 2, height: 9)
    }
}

/// Downward-pointing tip below the playhead grabber.
private struct Triangle: Shape {
    func path(in rect: CGRect) -> Path {
        var path = Path()
        path.move(to: CGPoint(x: rect.minX, y: rect.minY))
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.minY))
        path.addLine(to: CGPoint(x: rect.midX, y: rect.maxY))
        path.closeSubpath()
        return path
    }
}

File: android/app/src/main/kotlin/com/stillmoment/presentation/ui/meditations/trimeditor/TrimPlayheadGrabber.kt
package com.stillmoment.presentation.ui.meditations.trimeditor

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.unit.dp
import com.stillmoment.presentation.ui.theme.LocalStillMomentColors

/**
 * The sage playhead grabber, rendered as a full-size overlay over the waveform and
 * positioned in the upper (playhead) touch zone (shared-108). The full-height playhead
 * line is drawn by [TrimWaveformView]; this is only the grip with its downward tip.
 *
 * Purely visual — the actual dragging is resolved geometrically by the single track gesture
 * ([TrimHitTesting]: the upper 45 % belongs to the playhead). An off-window playhead shows
 * no grip (never glued to the edge). 1:1 port of iOS `TrimPlayheadGrabber`.
 */
@Composable
fun TrimPlayheadGrabber(
    playheadTimeMs: Long,
    window: ClosedRange<Long>,
    trackWidthPx: Float,
    waveformHeightPx: Float,
    modifier: Modifier = Modifier
) {
    val theme = LocalStillMomentColors.current
    val gradient = Brush.verticalGradient(listOf(theme.playheadAccentHi, theme.playheadAccent))
    Canvas(modifier = modifier.fillMaxSize()) {
        if (trackWidthPx <= 0f || !TrimGeometry.isTimeInWindow(playheadTimeMs, window)) {
            return@Canvas
        }
        val headX = TrimGeometry.x(playheadTimeMs, window, trackWidthPx)
        val zoneCenterY = waveformHeightPx * TrimHitTesting.VERTICAL_SPLIT / 2f
        drawGrabber(headX, zoneCenterY, gradient, theme.playheadAccent, theme.textOnPlayhead)
    }
}

private fun DrawScope.drawGrabber(
    headX: Float,
    zoneCenterY: Float,
    gradient: Brush,
    tipColor: Color,
    gripColor: Color
) {
    val grabberWidth = GRABBER_WIDTH_DP.dp.toPx()
    val grabberHeight = GRABBER_HEIGHT_DP.dp.toPx()
    val tipWidth = TIP_WIDTH_DP.dp.toPx()
    val tipHeight = TIP_HEIGHT_DP.dp.toPx()
    val totalHeight = grabberHeight + tipHeight
    val top = zoneCenterY - totalHeight / 2f

    drawRoundRect(
        brush = gradient,
        topLeft = Offset(headX - grabberWidth / 2f, top),
        size = Size(grabberWidth, grabberHeight),
        cornerRadius = CornerRadius(7.dp.toPx(), 7.dp.toPx())
    )

    // Two grip lines.
    val gripHeight = 9.dp.toPx()
    val gripWidth = 2.dp.toPx()
    val gripSpacing = 3.dp.toPx()
    val gripTop = top + grabberHeight / 2f - gripHeight / 2f
    listOf(headX - gripSpacing, headX + gripSpacing).forEach { gx ->
        drawRoundRect(
            color = gripColor.copy(alpha = 0.55f),
            topLeft = Offset(gx - gripWidth / 2f, gripTop),
            size = Size(gripWidth, gripHeight),
            cornerRadius = CornerRadius(gripWidth, gripWidth)
        )
    }

    // Downward-pointing tip below the grabber.
    val tipTop = top + grabberHeight
    val tipPath = Path().apply {
        moveTo(headX - tipWidth / 2f, tipTop)
        lineTo(headX + tipWidth / 2f, tipTop)
        lineTo(headX, tipTop + tipHeight)
        close()
    }
    drawPath(tipPath, color = tipColor)
}

private const val GRABBER_WIDTH_DP = 32f
private const val GRABBER_HEIGHT_DP = 20f
private const val TIP_WIDTH_DP = 10f
private const val TIP_HEIGHT_DP = 6f
